project/Settings.scala

```diff
@@ -22,11 +22,51 @@ object Settings {
   )
 
   val commonCompilerSettings = Seq(
-    "-unchecked",
-    "-deprecation",
-    "-feature",
-    "-encoding", "utf8",
     "-target:jvm-1.8",
+    "-deprecation",                      // Emit warning and location for usages of deprecated APIs.
+    "-encoding", "utf-8",                // Specify character encoding used by source files.
+    "-explaintypes",                     // Explain type errors in more detail.
+    "-feature",                          // Emit warning and location for usages of features that should be imported explicitly.
+    "-language:existentials",            // Existential types (besides wildcard types) can be written and inferred
+    "-language:higherKinds",             // Allow higher-kinded types
+    "-language:implicitConversions",     // Allow definition of implicit functions called views
+    "-unchecked",                        // Enable additional warnings where generated code depends on assumptions.
+    "-Xcheckinit",                       // Wrap field accessors to throw an exception on uninitialized access.
+    "-Xfatal-warnings",                  // Fail the compilation if there are any warnings.
```

Comment:
when it says _any_ warning does it really mean _any_ warning? i.e. all of the additional warnings that are being turned on in this PR will be grounds to fail the compilation?

Reply:
yep..with this option on, compilation will fail when there's warning.

Reply:
i think this is a good idea but we should communicate it to the team as it's a behavioral change that developers are likely to encounter